Hybrid topology

mtweiden opened the pull request:
Changed the `gen_successors` function to use topology. Still need to figure out how to take `MachineModel`s and split them into subtopologies.

ethanhs commented:
Maybe this PR should target the `iterative-reoptimization` branch as a base branch? Not sure how you want to handle that @edyounis . It seems that it would be easier to review this if it didn't have those changes as well.

mtweiden commented:
I've changed the target branch to `iterative-reoptimization` and fixed some pre-commit issues. Those that remain have to do with the `lang` files. I plan to make a few more commits to this PR that will add the ability to save checkpoints/intermediate results from synthesis.

ethanhs commented:
Excellent! Checkpointing is very useful

mtweiden commented:
Okay, I think this branch is ready to be merged with iterative-reoptimization if there are no more changes that y'all think need to happen.

edyounis reviewed (changes requested):
It seems that you are merging the native code PR into it as well. We should chat to figure out these branches before we merge this in. Otherwise, good work, just some small changes requested.

Nit: Please remove all `.DS_Store` files

ethanhs commented on .gitignore:
```diff
@@ -6,4 +6,5 @@
 .venv/
 .vscode/
 .mypy_cache/
+scratch/
```
You can always add a user specific ignore to avoid committing this folder if you wish: https://stackoverflow.com/a/5724499

mtweiden commented on bqskit/ir/circuit.py:
```diff
@@ -1547,11 +1548,11 @@ def surround(self, point: CircuitPointLike, size: int) -> CircuitRegion:
-            list[HalfWire],
-            list[tuple[int, Operation]],
-            CircuitLocation,
+        HalfWire = tuple[CircuitPoint, str]  # type: ignore
```
These were added because my interpreter doesn't like ` = tuple[~~]`. Is it okay if I change it to ` : tuple[~~]` instead?

edyounis replied:
I think, you just need to pull. Let's chat today if you have some time.

mtweiden commented on bqskit/compiler/passes/synthesispass.py:
```diff
@@ -97,18 +116,46 @@ def run(self, circuit: Circuit, data: dict[str, Any]) -> None:
                 new_utry = syn_circuit.get_unitary()
                 old_utry = op.get_unitary()
-                errors.append(new_utry.get_distance_from(old_utry))
+                error = new_utry.get_distance_from(old_utry)
```
Should I change this to `error = x**2` then?

edyounis replied:
No, it's a bit more than that. We can discuss this over slack or zoom. I'll send you a slack message.

mtweiden commented on bqskit/compiler/passes/synthesispass.py:
```diff
@@ -133,3 +186,10 @@ def default_collection_filter(op: Operation) -> bool:
+    circuit_copy = circuit.copy()
+    VariableToU3Pass().run(circuit_copy, {})
+    with open(path + f'/block_{num}', 'w') as f:
+        f.write(OPENQASM2Language().encode(circuit_copy))
```
I'm mostly using `SaveIntermediatePass` in `bqskit/compiler/passes/util/intermediate.py`, but that should probably get looked at again as well.

edyounis replied:
Yeah, then this is fine for now. Can you stick a TODO in there for now?
